compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs

```diff
@@ -0,0 +1,181 @@
+//! Move Array Gets Pass
+//!
+//! This is a setup optimization for the mutable array set optimization.
+//! Array sets can be made mutable if the array they're setting isn't used
+//! again afterward. This pass moves all array gets to their earliest possible
+//! location right after their creation. This pass is not expected to yield
+//! any runtime difference without the mutable array pass.
+//!
+//! To move an array get:
+//! - We must keep the enable_side_effects variable the same
+//! - It cannot move past any instructions it depends on. This includes:
+//!   - The array itself, the index to set, and the enable_side_effects variable.
+//!
+//! This optimization has two passes:
+//! 1. Traverse the function to find each array_get and remember its dependencies,
+//!    draining each instruction as we go.
+//! 2. Traverse the function re-inserting each instruction. When all dependencies
+//!    of an array_get have been inserted we can insert the corresponding array_get itself.
+//!    Since it can be expensive checking this for every instruction, this pass makes
+//!    the assumption that the last dependency will always have the highest ValueId.
+use fxhash::FxHashMap;
```

Comment:
```suggestion
use fxhash::FxHashMap as HashMap;
```
nit: just for consistency with what we do throughout the codebase